roofit/roostats/src/RooStatsUtils.cxx

```diff
@@ -93,23 +93,19 @@ namespace RooStats {
    void FactorizePdf(const RooArgSet &observables, RooAbsPdf &pdf, RooArgList &obsTerms, RooArgList &constraints) {
    // utility function to factorize constraint terms from a pdf
    // (from G. Petrucciani)
-      const std::type_info & id = typeid(pdf);
-      if (id == typeid(RooProdPdf)) {
-         RooProdPdf *prod = dynamic_cast<RooProdPdf *>(&pdf);
+      if (auto prod = dynamic_cast<RooProdPdf *>(&pdf)) {
          RooArgList list(prod->pdfList());
          for (int i = 0, n = list.getSize(); i < n; ++i) {
             RooAbsPdf *pdfi = (RooAbsPdf *) list.at(i);
             FactorizePdf(observables, *pdfi, obsTerms, constraints);
          }
-      } else if (id == typeid(RooExtendPdf)) {
-         TIterator *iter = pdf.serverIterator();
+      } else if (dynamic_cast<RooExtendPdf *>(&pdf)) {
          // extract underlying pdf which is extended; first server is the pdf; second server is the number of events variable
-         RooAbsPdf *updf = dynamic_cast<RooAbsPdf *>(iter->Next());
-         assert(updf != 0);
-         delete iter;
-         FactorizePdf(observables, *updf, obsTerms, constraints);
-      } else if (id == typeid(RooSimultaneous)) {    //|| id == typeid(RooSimultaneousOpt)) {
-         RooSimultaneous *sim  = dynamic_cast<RooSimultaneous *>(&pdf);
+         auto iter = pdf.servers().begin();
+         assert(iter != pdf.servers().end());
+         assert(dynamic_cast<RooAbsPdf*>(*iter));
+         FactorizePdf(observables, static_cast<RooAbsPdf&>(**iter), obsTerms, constraints);
```

Comment:
```diff
-      auto iter = pdf.servers().begin();
-      assert(iter != pdf.servers().end());
-      assert(dynamic_cast<RooAbsPdf*>(*iter));
-      FactorizePdf(observables, static_cast<RooAbsPdf&>(**iter), obsTerms, constraints);
+      assert(!pdf.servers().empty() && dynamic_cast<RooAbsPdf*>(pdf.servers().first()));
+      FactorizePdf(observables, static_cast<RooAbsPdf&>(*pdf.servers().first()), obsTerms, constraints);
```
I would even suggest to implement (== forward the call to the underlying vector) `RooSTLRefCountList::front()` and `back()`. The goal was to make it more STL-like anyway.